xl: Avoid removing directory content in Delete API

vadmeste opened the pull request:
## Description
Delete & Multi Delete API should not try to remove the directory content.
The only permitted case is with zero size object with a trailing slash
in its name.

## Motivation and Context
#5547 

## How Has This Been Tested?
go test + Manually

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

harshavardhana reviewed (dismissed):
LGTM tested

poornas reviewed (changes requested):
@vadmeste, while this change prevents removing non empty directories, it is also preventing deleting empty directories. 

```
curl -X PUT localhost:9000/imags/em3/temp3/
mc policy upload myminio/imags
curl -X DELETE http://localhost:9000/imags/em3/temp3/
➜  testpgm mc ls myminio/imags                                  
[2018-02-20 10:37:53 PST]     0B em2/
[2018-02-20 10:37:53 PST]     0B em3/
[2018-02-20 10:37:53 PST]     0B empty/

```

harshavardhana commented on cmd/xl-v1-object.go:
```diff
@@ -774,21 +774,16 @@ func (xl xlObjects) putObject(bucket string, object string, data *hash.Reader, m
-	// If its a directory request, no need to read metadata.
-	if !hasSuffix(object, slashSeparator) {
-		// Read metadata associated with the object from all disks.
-		metaArr, errs := readAllXLMetadata(xl.getDisks(), bucket, object)
```
is this change still needed?

vadmeste replied:
> is this change still needed?

This is a non functional change, I just wanted to separate removing object & removing directory

harshavardhana replied:
My suggestion is to avoid that in this PR @vadmeste

vadmeste replied:
> My suggestion is to avoid that in this PR @vadmeste

I also wanted to send a more minimal change but it turned out this is the simplest way to do that without being controversial.

harshavardhana replied:
we already know we support directory operations cleanup is not needed IMHO simply fix the bug you can do cleanup later as well.

poornas commented on cmd/xl-v1-object.go:
```diff
@@ -815,6 +810,37 @@ func (xl xlObjects) deleteObject(bucket, object string) error {
 	return reduceWriteQuorumErrs(dErrs, objectOpIgnoredErrs, writeQuorum)
 }
 
```
function needs a comment.

vadmeste committed:
xl: Avoid removing directory content in Delete API

Delete & Multi Delete API should not try to remove the directory content.
The only permitted case is with zero size object with a trailing slash
in its name.

poornas reviewed (approved):
Tested. LGTM
